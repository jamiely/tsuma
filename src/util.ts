import { Color, Point, Rectangle } from "./types";

const colors: Color[] = ["red", "green", "blue", "gold", "purple"];
export const randomColor = () =>
  colors[Math.floor(Math.random() * colors.length)];

export const distance = (pt1: Point, pt2: Point) => {
  const dx = pt1.x - pt2.x;
  const dy = pt1.y - pt2.y;
  return Math.sqrt(dx * dx + dy * dy);
};

export const inBounds = (pt: Point, {position, size}: Rectangle) => {
  if(pt.x < position.x) return false;
  if(pt.x > position.x + size.width) return false;
  if(pt.y < position.y) return false;
  if(pt.y > position.y + size.height) return false;

  return true;
}

export const magnitude = ({ x, y }: Point) => Math.sqrt(x * x + y * y);
export const add = (a: Point, b: Point) => {
  a.x += b.x;
  a.y += b.y;
};
export const subtract = (a: Point, b: Point) => {
  a.x -= b.x;
  a.y -= b.y;
};
export const scale = (a: Point, factor: number) => {
  a.x *= factor;
  a.y *= factor;
};

export const toUnit = (pt: Point) => {
  const mag = magnitude(pt);
  pt.x /= mag;
  pt.y /= mag;
};

// via chat GPT
export function getIntersection(
  A: Point, B: Point,
  C: Point, D: Point
): Point | null {
  // Line AB represented as a1x + b1y = c1
  const a1 = B.y - A.y;
  const b1 = A.x - B.x;
  const c1 = a1 * A.x + b1 * A.y;

  // Line CD represented as a2x + b2y = c2
  const a2 = D.y - C.y;
  const b2 = C.x - D.x;
  const c2 = a2 * C.x + b2 * C.y;

  const determinant = a1 * b2 - a2 * b1;

  if (determinant === 0) {
      // The lines are parallel, so no intersection
      return null;
  } else {
      const x = (b2 * c1 - b1 * c2) / determinant;
      const y = (a1 * c2 - a2 * c1) / determinant;

      return { x, y };
  }
}
