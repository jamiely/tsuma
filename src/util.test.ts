import { expect, test } from 'vitest'
import { add, distance, getIntersection, inBounds, magnitude, scale, setPoint, subtract, toUnit } from './util'

test('distance', () => {
  expect(distance({x: 0, y:0}, {x: 2, y: 3})).toBe(Math.sqrt(13));
})  

test('inBounds', () => {
  const bounds = {position: {x: 1, y: 1}, size: {width: 4, height: 9}}
  expect(inBounds({x: 1, y: 1}, bounds)).toBe(true);
  expect(inBounds({x: 5, y: 10}, bounds)).toBe(true);
  expect(inBounds({x: 0, y: 1}, bounds)).toBe(false);
  expect(inBounds({x: 1, y: 0}, bounds)).toBe(false);
  expect(inBounds({x: 6, y: 10}, bounds)).toBe(false);
  expect(inBounds({x: 5, y: 11}, bounds)).toBe(false);
})

test('magnitude', () => {
  expect(magnitude({x: 1, y: 2})).toBe(Math.sqrt(5));
})

test('add', () => {
  const pt = {x: 1, y: 1};
  add(pt, {x: 3, y: 4});
  expect(pt).toStrictEqual({x: 4, y: 5});
})

test('subtract', () => {
  const pt = {x: 3, y: 4};
  subtract(pt, {x: 1, y: 1});
  expect(pt).toStrictEqual({x: 2, y: 3});
})

test('scale', () => {
  const pt = {x: 2, y: 3};
  scale(pt, 2);
  expect(pt).toStrictEqual({x: 4, y: 6});
})

test('toUnit', () => {
  const pt = {x: 2, y: 3};
  toUnit(pt);
  expect(pt.x - 0.5547001962252291).toBeLessThan(0.0001);
  expect(pt.y - 0.8320502943378437).toBeLessThan(0.0001);
})

test('setPoint', () => {
  const pt = {x: 1, y: 1};
  setPoint(pt, {x: 2, y: 3});
  expect(pt).toStrictEqual({x: 2, y: 3});
});

test('getIntersection', () => {
  const intersection = getIntersection({x: 1, y:1}, {x: 5, y: 5}, {x: -1, y: -1}, {x: -5, y: 2});
  expect(intersection).toStrictEqual({x: -1, y: -1});
})
