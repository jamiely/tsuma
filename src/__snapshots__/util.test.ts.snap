// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`chaikinSmoothing 1`] = `
[
  {
    "x": 0.4375,
    "y": 1,
  },
  {
    "x": 0.5625,
    "y": 1,
  },
  {
    "x": 0.671875,
    "y": 0.984375,
  },
  {
    "x": 0.765625,
    "y": 0.953125,
  },
  {
    "x": 0.84375,
    "y": 0.90625,
  },
  {
    "x": 0.90625,
    "y": 0.84375,
  },
  {
    "x": 0.953125,
    "y": 0.765625,
  },
  {
    "x": 0.984375,
    "y": 0.671875,
  },
  {
    "x": 1,
    "y": 0.5625,
  },
  {
    "x": 1,
    "y": 0.4375,
  },
]
`;
